src/fileimpl.cpp

```diff
@@ -400,23 +400,18 @@ namespace zim
     if (idx >= getCountClusters())
       throw ZimFileFormatError("cluster index out of range");
 
+    std::promise<ClusterHandle> clusterPromise;
     pthread_mutex_lock(&clusterCacheLock);
-    auto cluster(clusterCache.get(idx));
+    const auto x = clusterCache.put(idx, clusterPromise.get_future().share());
     pthread_mutex_unlock(&clusterCacheLock);
-    if (cluster)
-    {
+    if ( !x.first ) {
```

Comment:
Do we really need to create the promise at this level ?
We use the promise only to create a shared_future and (potentially) add it to the cache.

We could make the cache create itself the shared_future if needed.
Thus :
- The `clusterCacheLock` can be move internally to the cache (and handle by it). 
- The whole "main interaction" with the cache would simply be `const auto x = clusterCache.get(idx);` (x keeping the same type).

Reply:
> Do we really need to create the promise at this level ?
> We use the promise only to create a shared_future and (potentially) add it to the cache.

The promise is used to the set the value of the new entry in case of a cache miss. But of course it can be hidden inside a higher-level thread-safe cache with the API that you proposed.